source/hwIo/ioThread.py

```diff
@@ -39,149 +52,190 @@ class IoThread(threading.Thread):
 	"""
 
 	exit: bool = False
-	_apcReferences: typing.Dict[uuid.UUID, winKernel.PAPCFUNC]
+	_apcStore: ApcStoreT
+	_completionRoutineStore: CompletionRoutineStoreTypeT
 
 	def __init__(self):
 		super().__init__(
 			name=f"{self.__class__.__module__}.{self.__class__.__qualname__}",
 			daemon=True
 		)
-		self._apcReferences = dict()
+		self._apcParamCounter = self._generateApcParams()
+		self._internalApcInstance = winKernel.PAPCFUNC(self._internalApc)
+		self._internalCompletionRoutineInstance = LPOVERLAPPED_COMPLETION_ROUTINE(self._internalCompletionRoutine)
```

Comment:
Note that this creates a circular reference. It's also theoretically possible this could be called after death if the IoThread instance is destroyed before the callback fires.

Both of these things are fine if you are absolutely certain the IoThread instance will outlive any OVERLAPPED or APC calls. I'm pretty sure it always will, since nothing can fire outside of an alertable wait and we join the thread when we exit it. Still, I thought it worth noting.

Reply:
FWIW, for WasapiWavePlayer, the ctypes callback is a method on the class rather than the instance. I maintain a dict on the class which maps C++ WasapiPlayer pointers to Python instances. You might be able to do something similar, mapping thread ids to instances or something like that. However, it might not be worth it if you're sure we're meeting the above guarantees.

Reply:
In this case, better safe than sorry, so I think I'll follow your proposal.